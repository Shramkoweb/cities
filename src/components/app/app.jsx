import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import MainPage from "../main-page/main-page";

const App = (props) => {
  const {offers} = props;

  return (
    <MainPage
      offers={offers}
    />
  );
};

App.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.object).isRequired
};

const mapStateToProps = (state) => Object.assign({}, {
  city: state.city,
});

export {App};
export default connect(mapStateToProps)(App);
