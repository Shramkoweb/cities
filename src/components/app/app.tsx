import * as React from "react";
import {connect} from "react-redux";
import {Redirect, Route, Switch} from "react-router-dom";
import MainPage from "../main-page/main-page";
import {getAuthorizationStatus} from "../../reducer/user/selector";
import Sign from "../sign/sign";
import {PageAddress} from "../../constants";
import Loader from "../loader/loader";
import Property from "../property/property";
import {getError, getLoadingStatus} from "../../reducer/data/selector";
import withPrivateRoute from "../../hocs/with-private-route/with-private-route";
import FavoritesPage from "../favorites-page/favorites-page";
import {Operation} from "../../reducer/user/user";
import withExistOffer from "../../hocs/with-exist-offer/with-exist-offer";

const PropertyWrapped = withExistOffer(Property);

interface AppProps {
  onCheckAuth: () => void,
  isLoading: boolean,
  isAuthorizationRequired: boolean,
  isError: string,
}

class App extends React.PureComponent<AppProps> {
  componentDidMount() {
    const {onCheckAuth} = this.props;

    onCheckAuth();
  }

  render() {
    const {isLoading, isAuthorizationRequired, isError} = this.props;
    const FavoritesPrivate = withPrivateRoute(isAuthorizationRequired, PageAddress.LOGIN)(FavoritesPage);
    const SignPrivate = withPrivateRoute(!isAuthorizationRequired)(Sign);

    return (
      <>
        {
          isLoading ? <Loader isError={isError}/>
            :
            <Switch>
              <Route path={PageAddress.MAIN} exact component={MainPage}/>
              <Route path={PageAddress.LOGIN} exact component={SignPrivate}/>
              <Route path={`${PageAddress.OFFER}/:id`} exact render={({match}) =>
                <PropertyWrapped id={Number(match.params.id)}/>}/>
              <Route path={PageAddress.FAVORITE} exact component={FavoritesPrivate}/>
              <Redirect from='*' to={PageAddress.MAIN}/>
            </Switch>
        }
      </>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  onCheckAuth: () => dispatch(Operation.onCheckAuth()),
});

const mapStateToProps = (state) => ({
  isAuthorizationRequired: getAuthorizationStatus(state),
  isLoading: getLoadingStatus(state),
  isError: getError(state)
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
