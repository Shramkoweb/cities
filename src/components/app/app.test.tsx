import * as React from "react";
import * as Enzyme from "enzyme";
import {render} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

import {App} from "./app";

jest.mock(`../../index.tsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../../index.tsx`, () => jest.fn().mockReturnValue(null));

Enzyme.configure({adapter: new Adapter()});

it(`Is App render`, () => {
  const tree = render((<App
    isAuthorizationRequired={false}
    isLoading={true}
    onCheckAuth={jest.fn()}
    isError={``}
  />));

  expect(toJson(tree)).toMatchSnapshot();
});
