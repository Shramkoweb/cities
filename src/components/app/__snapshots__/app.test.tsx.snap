// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Is App render 1`] = `
<div
  class="loader-wrapper"
>
  <p
    class="loader"
  >
    Loading...
  </p>
</div>
`;
